### PR TITLE
WIP: replace flow with typescript-jsdoc

### PR DESCRIPTION
**Pull request by vankop**
> Which issue, if any, is this issue related to?

https://github.com/stylelint/stylelint/issues/4245

Help wanted 🤗

### CODE REVIEW

**Comment by hudochenkov** [comment]
I see you removed flow directive from all files. You decided to remove all Flow code first, and add JSDoc whenever you can think of? Or Flow is still running?

Could you add TypeScript to npm scripts, please? Can it also be added to `pretest` npm script so we run TypeScript and Flow together?

**Comment by vankop** [comment]
Flow is running, but does not lint anything, I did it just for development purpose (don’t think about flow types). We can revert this commit and commit that removes flow-typed later

**Comment by hudochenkov** [comment]
Looks like our actual JavaScript got broken in the last commit.

**Comment by vankop** [comment]
> Looks like our actual JavaScript got broken in the last commit.

yes you are correct =) I will revert this for now

**Comment by vankop** [comment]
If someone wants to help I suggest to write what you are taking here.

I will fix `lib/assignDisabledRanges.js`
and take
```
lib/standalone.js
lib/writeOutputFile.js
```

I think files from [this list](https://github.com/stylelint/stylelint/issues/4245#issuecomment-536902723) are prioritize

**Comment by vankop** [comment]
I have created issue related to `~/lib/assignDisabledRanges` https://github.com/stylelint/stylelint/issues/4328

**Comment by vankop** [comment]
@hudochenkov I think main work is done =) so I added types check in lint stage. Mostly need to write types for `utils` and fix `TODO TYPES`

**Comment by hudochenkov** [comment]
Does TS failing because it's a WIP?

**Comment by vankop** [comment]
Output looks a little bit strange, but looks like it works. It fails because it is around 200 warnings still . I did not found out why it is linting some files that are in excluded patterns =(

**Comment by vankop** [comment]
I have fixed excluded patterns, now warnings (~50 at all) are more readable =)

File: types/meow/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+interface Flag {
```
Comment (G-Rath) [inline types/meow/index.d.ts]: There is an [`@types/meow`](https://www.npmjs.com/package/@types/meow) that looks pretty similar to this, so you should be able to just use that.

File: types/global-modules/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+declare module 'global-modules' {
```
Comment (G-Rath) [inline types/global-modules/index.d.ts]: There is an [`@types/global-modules`](https://www.npmjs.com/package/@types/global-modules) that looks pretty similar to this, so you should be able to just use that.

**Comment by G-Rath** [comment]
I've added a few comments about `.d.ts` files that can be replaced w/ their `@types` package.

If there is a library that doesn't have a type definition file published on DefinitelyTyped, I'm happy to create one, and get it published.

**Comment by vankop** [comment]
@G-Rath Do you now any workaround of using `default` definitions. What I mean, for example for `strip-ansi` there are definitions in format https://github.com/chalk/strip-ansi/blob/master/index.d.ts, but they don't provide es6 exports (as defined in types). So maybe there are some workaround?

**Comment by G-Rath** [comment]
@vankop `export default` is ES6-style exporting, so you can use ES6 style importing:

```
import blah from 'strip-ansi';
```

**Comment by vankop** [comment]
Yes, you are correct for ES6, but we have code in CommonJS and `strip-ansi` also provide CommonJS export only (but types provide ES6 export). See

> for `strip-ansi` there are definitions in format https://github.com/chalk/strip-ansi/blob/master/index.d.ts, but they don't provide es6 exports

**Comment by G-Rath** [comment]
Ah yes, sorry now I'm w/ you - the commit you pushed is the workaround; that's what `esModuleInterop` does: it transpiles everything to be exported as `{ default: ... }` to fix that issue.

Since you're in JS land however, you've got to do it yourself.

**Comment by G-Rath** [comment]
FYI I've created a PR for adding definitions for [`globjoin`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38908). They should be merged & published within the week :)

**Comment by vankop** [comment]
@G-Rath I know how it works =) I just did not realizes that so many packages support both ways. Thanks for deep explanation! (babel as well provides fallback for both exports `_interopRequireDefault`)

**Comment by vankop** [comment]
@G-Rath As I know there is no a lot of typescript in jsdoc examples + it poor documented (if compare with common Typescript), could you suggest any source of examples / rich information about typescript in jsdoc? as I know there are a lot of tests here - https://github.com/microsoft/TypeScript/tree/master/tests/cases/fourslash. However, it hard to find usage some use cases there (folder is pretty large)

**Comment by G-Rath** [comment]
@vankop sadly not really; you've pretty much summed up the state of the situation.

The two places I'd recommend are looking at [`tsdoc`](https://github.com/microsoft/tsdoc) - It's a project so underway it's not been announced, but I stumbled upon it myself a few months ago when looking for info like you. That can give you a good idea of at least how MS/TS would *like* JS/TSDocs to endup looking - if you just remember to ignore tags that are not official JSDoc ones, I'd expect their examples to already work in TS.

The other place of information is [the handbook](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html).

Finally, the eslint team are actually looking into this same thing, and so you might find interesting code in the [PR](https://github.com/eslint/eslint/pull/12082).

**Comment by vankop** [comment]
@hudochenkov looks like problem with CI still exists

File: lib/cli.js [hunk 3]
```diff
@@ -505,37 +429,44 @@ module.exports = (argv /*: string[]*/) /*: Promise<void>|void*/ => {
...
-	if (cli.flags.help || cli.flags.h) {
+	if (cli.flags.help) {
 		cli.showHelp(0);
 
-		return;
+		return Promise.resolve();
 	}
 
-	if (cli.flags.version || cli.flags.v) {
+	if (cli.flags.version) {
```
Comment (hudochenkov) [inline lib/cli.js]: `-h` and `-v` removed. Please, add them back.

File: lib/createStylelintResult.js [hunk 4]
```diff
@@ -1,87 +1,25 @@
...
+ * @return {Promise<StylelintResult>}
+ */
+module.exports = function(stylelint, postcssResult, filePath, cssSyntaxError) {
+	/** @type {StylelintResult} */
 	let stylelintResult;
+	/** @type {string | undefined} */
 	let source;
 
-	if (postcssResult) {
+	if (postcssResult && postcssResult.root) {
```
Comment (hudochenkov) [inline lib/createStylelintResult.js]: Why add this check?

File: lib/getPostcssResult.js [hunk 5]
```diff
@@ -1,29 +1,27 @@
...
+/**
+ * @param {import('stylelint').StylelintInternalApi} stylelint
+ * @returns {Promise<import('postcss').Result>}
+ */
+module.exports = function(stylelint) {
+	/** @type {import('stylelint').GetPostcssOptions} */
+	const options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
 
-	const cached /*: ?postcss$result*/ = stylelint._postcssResultCache.get(options.filePath);
+	const cached = options.filePath ? stylelint._postcssResultCache.get(options.filePath) : undefined;
```
Comment (hudochenkov) [inline lib/getPostcssResult.js]: Why this ternary?

File: lib/getPostcssResult.js [hunk 6]
```diff
@@ -58,7 +58,7 @@ module.exports = function(stylelint /*: stylelint$internalApi*/) /*: Promise<?Ob
 						stringify: postcss.stringify,
 					};
 				}
-			} else if (syntax) {
+			} else if (stylelint._options.syntax) {
```
Comment (hudochenkov) [inline lib/getPostcssResult.js]: Why not use `syntax` as before?

File: lib/getPostcssResult.js [hunk 7]
```diff
@@ -122,13 +122,23 @@ module.exports = function(stylelint /*: stylelint$internalApi*/) /*: Promise<?Ob
 			return result;
 		})
 		.then((postcssResult) => {
-			stylelint._postcssResultCache.set(options.filePath, postcssResult);
+			if (options.filePath) {
```
Comment (hudochenkov) [inline lib/getPostcssResult.js]: Why additional check?

**Comment by hudochenkov** [comment]
In a totally new places. Might be that it happens with code changes made in this PR. Though locally nothing fails in this branch.

File: lib/createStylelintResult.js (hunk 4 above)
Comment (vankop) [reply lib/createStylelintResult.js]: Root possible undefined

File: lib/getPostcssResult.js (hunk 5 above)
Comment (vankop) [reply lib/getPostcssResult.js]: You can’t pass undefined in Map get parameter

File: lib/getPostcssResult.js (hunk 7 above)
Comment (vankop) [reply lib/getPostcssResult.js]: options.filePath possible undefined same with `get` method

File: lib/getPostcssResult.js (hunk 6 above)
Comment (vankop) [reply lib/getPostcssResult.js]: here is a little bit tricky place `stylelint._options.syntax` is `undefined | string` so after assigning on line 41 `syntax` implicitly have type `undefined | string`, but after we assign to `syntax` object - `{parse: any, stringify: any}` as well.

**Comment by vankop** [comment]
> This PR would be close to impossible to review :(

Obviously, it is a problematic to review, if you have an idea how to split this, I will be glad to discuss.

> I don't think it makes much sense to do this comparison before this PR is ready

Sorry, I did not have time to fix last 20 warnings. All welcome to take a look on them =)

**Comment by G-Rath** [comment]
> if you have an idea how to split this, I will be glad to discuss.

I would recommend focusing on adding the types, rather than removing flow.

For example, there are a lot of files that have been changed b/c you're removed `@flow` from the top - but there's no reason to do that in this PR, since they're just comments.

Even if flow is generating errors, the whole point of the PR is to move away from flow, so the errors shouldn't matter.

You can also remove the typing files for flow in another PR, as while sure they're not needed anymore, they're also not doing anything and so don't *need* to be removed in this PR :)

**Comment by ntwb** [comment]
Thanks @G-Rath, and thanks for all your work thus far @vankop, this is all so much out of my wheelhouse, and anything that will help make reviewing this PR easier I'm all for.

So, if we can get away without including the removal of Flow and `@flow` comments that would be great if possible please.

**Comment by vankop** [comment]
First I can suggest create separate PR with tsconfig and provided types. I think we can take types folder from this PR, then we can make PRs step by step with or without flow (on your demand).

> I would recommend focusing on adding the types, rather than removing flow.

I did it for the reason as described above - just for development reasons. As I see now it is not necessary (however some times flow and ts types are misleading)

> You can also remove the typing files for flow in another PR, as while sure they're not needed anymore, they're also not doing anything and so don't _need_ to be removed in this PR :)

Yes you are right, we can revert this commit

> Even if flow is generating errors, the whole point of the PR is to move away from flow, so the errors shouldn't matter.

I think the main idea to stay with flow while TS will not cover stylelint' "core"

**Comment by vankop** [comment]
@hudochenkov after restarting CI still fail for Node.js 12.x.x, maybe it is relates to my changes =)

**Comment by ntwb** [comment]
> after restarting CI still fail for Node.js 12.x.x, maybe it is relates to my changes =)

I cannot get the logs of the failing jobs to display past line `29` 😞

**Comment by hudochenkov** [comment]
@ntwb it's a scrollable area. I always keep scrollbar enabled in OS settings :)

**Comment by vankop** [comment]
types is fine now, I think I will start to split this in several PRs

**Comment by G-Rath** [comment]
@vankop btw the definitions for `globjoin` got published :)

I'll circle back during the week and look to create definitions for the other libraries when I have a chance.

**Comment by vankop** [comment]
Thanks all for your reviews! All this work is done in separate issues 🎉

**Pull request closed without merging**